src/core/processing/qgsprocessingparameters.h

```diff
@@ -1245,6 +1247,20 @@ class CORE_EXPORT QgsProcessingParameters
      */
     static QgsCoordinateReferenceSystem parameterAsPointCrs( const QgsProcessingParameterDefinition *definition, const QVariant &value, QgsProcessingContext &context );
 
+    /**
+     * Evaluates the parameter with matching \a definition to a geometry.
+     *
+     * \since QGIS 3.16
+     */
+    static QgsGeometry parameterAsGeometry( const QgsProcessingParameterDefinition *definition, const QVariantMap &parameters, QgsProcessingContext &context );
+
+    /**
+     * Evaluates the parameter with matching \a definition and \a value to a geometry.
+     *
+     * \since QGIS 3.16
+     */
+    static QgsGeometry parameterAsGeometry( const QgsProcessingParameterDefinition *definition, const QVariant &value, QgsProcessingContext &context );
```

Comment:
There's an important consideration missing here. If you checkout the corresponding methods for the extent or point parameter types, you'll see that they accept a QgsCoordinateReferenceSystem argument, and that there's also corresponding "parameterAsPointCrs"/"parameterAsExtentCrs" methods. 

This is made possible because the extent/point parameter types handle (and prefer) QgsReferencedPointXY and QgsReferencedRectangle values, which include source CRS information along with the point/rectangle. It's what ultimately allows algorithms to gracefully handle the situation where the point/extent is in a different CRS vs the layer they are being used against (e.g. "clip by extent" -> draw extent on canvas in EPSG3857, clipped against a layer in EPSG4326). If we don't do this, then we're sliding back to the processing QGIS 2.x behavior were users had to manually ensure all inputs to an algorithm were in the same CRS.

So you'll need to:
1. Implement QgsReferencedGeometry (see QgsReferencedRectangle for how to do this, it's straightforward). Add tests for this new class.
2. Register the metatype for QgsReferencedGeometry so it can be stored in variants
3. Add the destination crs argument to parameterAsGeometry, and handle the transformation of geometries in parameterAsGeometry if the crs is specified (again, copy the parameterAsPoint/parameterAsExtent code for this) (with tests)
4. Add corresponding "parameterAsGeometryCrs" in order to retrieve the baked-in crs associated with a geometry parameter value (with tests)

Reply:
Yes, the lack of 'refererenced' geometry has bothered me at some point, but I couldn't decide about the relevancy of implementing it or not, now I got the point.

Reply:
@nyalldawson 

For QgsReferencedPoint and QgsReferencedExtent  the string representation append the crs definition enclosed  in square brackets [ ].  This is an adhoc representation for easy parsing. 

For WKT representation there is a [canonical form used by postgis - EWKT](https://postgis.net/docs/using_postgis_dbmanagement.html#EWKB_EWKT) , quite easier to parse with WKT geometry representation. AMHA, this representation seems more suited for string representation of CRS is paramater values. What do you think ?

Reply:
EWKT sounds ideal, if it's not too much work for you to handle this!

Reply:
> EWKT sounds ideal, if it's not too much work for you to handle this!

No, quite easy in fact, as it is an unambiguous prefix.  It's already done in the PR (with tests)